core-blocks/embed/index.js

```diff
@@ -46,6 +40,211 @@ const findBlock = ( url ) => {
 	return 'core/embed';
 };
 
+export function getEmbedEdit( title, icon ) {
+	return class extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.switchBackToURLInput = this.switchBackToURLInput.bind( this );
+			this.setUrl = this.setUrl.bind( this );
+			this.processPreview = this.processPreview.bind( this );
+
+			this.state = {
+				editingURL: false,
+				url: this.props.attributes.url,
+			};
+
+			if ( this.props.preview ) {
+				this.processPreview();
+			}
+		}
+
+		componentWillUnmount() {
+			// can't abort the fetch promise, so let it know we will unmount
+			this.unmounting = true;
+		}
+
+		componentDidUpdate( prevProps ) {
+			const hasPreview = undefined !== this.props.preview;
+			const hadPreview = undefined !== prevProps.preview;
+			// We had a preview, and the URL was edited, and the new URL already has a preview fetched.
+			const switchedPreview = this.props.preview && this.props.attributes.url !== prevProps.attributes.url;
+			if ( ( hasPreview && ! hadPreview ) || switchedPreview ) {
+				this.processPreview();
+			}
+		}
+
+		getPhotoHtml( photo ) {
+			// 100% width for the preview so it fits nicely into the document, some "thumbnails" are
+			// acually the full size photo.
+			const photoPreview = <p><img src={ photo.thumbnail_url } alt={ photo.title } width="100%" /></p>;
+			return renderToString( photoPreview );
+		}
+
+		setUrl( event ) {
+			if ( event ) {
+				event.preventDefault();
+			}
+			const { url } = this.state;
+			const { setAttributes } = this.props;
+			this.setState( { editingURL: false } );
+			if ( url === this.props.attributes.url ) {
+				if ( this.props.preview ) {
+					// Form has been submitted without changing the url, and we already have a preview,
+					// so process it again. Happens when the user clicks edit and then immediately
+					// clicks embed.
+					this.processPreview();
```

Comment:
Do we really need to to do anything here? since the url stayed the same, why are we calling `processPreview` here?

Reply:
This is one of the areas I need to clean up, because so much of the state isn't there any more. I'll mark this PR as WIP until I have everything cleaned up.